### DISCUSSION TITLE
Enumerable#lazy, TCPSocket, thread safety, etc.

### DISCUSSION THREAD

**Discussion by PhilCoggins** [ioquatix/ioquatix#7] [General]
Hello Samuel 👋 . Your name came up on a thread in https://github.com/httprb/http/issues/692, so I found your user page and that you are open to discussions surrounding Ruby and concurrency and wanted to see if you were interested in looking at the issue and maybe try to enlighten me. I'm failing to provide a reproducible test case for TCPSocket.open getting hung in some circumstances. The issue seems to have surfaced after a recent change that attempts to open N + (Number of A records returned from `Resolv.getaddresses(host_name)`), as opposed to opening one single connection.

The connect timeout uses `Timeout.timeout`, of which I am aware of the various issues, but I don't quite understand in what circumstances it could cause TCPSocket.open to hang indefinitely. My organization uses Sidekiq and thus we have a lot of concurrent work, and any time these issues crop up I'm generally the one to step in and figure out the core issue. Any troubleshooting steps or clues to get this to reproduce would be very much appreciated. Thanks for your time and your contributions, take care!

**Comment by ioquatix** [comment]
Do you have any links to the change:

> The issue seems to have surfaced after a recent change that attempts to open N + (Number of A records returned from Resolv.getaddresses(host_name)), as opposed to opening one single connection.

**Reply by PhilCoggins** [reply]
Yes, it is here https://github.com/httprb/http/pull/638/

**Reply by ioquatix** [reply]
`Timeout.timeout` uses brute force to interrupt a system call. Without at least a basic exception safety, this can cause internal state issues.

It should be possible to reproduce this in a small example, i.e. a loop which does 10,000 requests and a timeout which kills some of them. Eventually you'd expect to reproduce the stuck state.

I'm happy to help with this but if you think you are up to it, why don't you have a go at trying to create a repro script?

**Reply by PhilCoggins** [reply]
Thanks! I have worked on and off with varying repro scripts, I'll paste that in below.

I am opening sockets to `93.184.216.34` (example.com). Setting the `Timeout` properly is tricky - I want to at least give enough time to open the socket with the OS and let it be in flight, but not to fully succeed as so it never times out. I ran a basic Benchmark.measure about 10 times and found a sweet spot at `0.015`, this will likely be dependent on hardware, OS or other factors.
```ruby
Benchmark.measure { TCPSocket.open("93.184.216.34", "443") }
```

I also implemented a short sleep, 1. because this more closely simulates real-world and 2. because once it hits a certain threshold, about 450 iterations on my machine, I get the below endlessly:
```
#<Thread:0x00007fe4e7204058 repro_socket_hang.rb:13 run> terminated with exception (report_on_exception is true):
repro_socket_hang.rb:14:in `initialize': Too many open files - socket(2) for "93.184.216.34" port 443 (Errno::EMFILE)
	from repro_socket_hang.rb:14:in `open'
	from repro_socket_hang.rb:14:in `block (3 levels) in <main>'
	from /Users/philcoggins/.rbenv/versions/3.0.1/lib/ruby/3.0.0/timeout.rb:97:in `block in timeout'
	from /Users/philcoggins/.rbenv/versions/3.0.1/lib/ruby/3.0.0/timeout.rb:35:in `block in catch'
	from /Users/philcoggins/.rbenv/versions/3.0.1/lib/ruby/3.0.0/timeout.rb:35:in `catch'
	from /Users/philcoggins/.rbenv/versions/3.0.1/lib/ruby/3.0.0/timeout.rb:35:in `catch'
	from /Users/philcoggins/.rbenv/versions/3.0.1/lib/ruby/3.0.0/timeout.rb:112:in `timeout'
	from repro_socket_hang.rb:14:in `block (2 levels) in <main>'

# man 2 socket:
# [EMFILE]           The per-process descriptor table is full.
```

I'll continue to research, but so far have been unable to get anything to hang endlessly. Thanks for looking!

```ruby
#!/usr/bin/env ruby

require 'socket'
require 'timeout'

iterations = (ENV["ITERATIONS"] || 100).to_i
threads = []
sockets = {}

begin
  iterations.times do |i|
    sleep(0.05)
    threads << Thread.new do
      Timeout.timeout(0.015) { sockets[i] = TCPSocket.open("93.184.216.34", "443") }
    end
  end

  threads.map(&:join)
ensure
  puts "### Threads"
  puts threads

  puts "### Sockets"
  puts sockets.map { |idx, sock| "#{idx} - addr: #{sock.addr} peeraddr: #{sock.peeraddr}" }

  puts "Thread count: #{threads.size}"
  puts "Socket count: #{sockets.size}"
end
```
